### PR TITLE
Patched docs for torch_compile_tutorial

### LINKED ISSUE #2858: Better specify `torch.compile behaviour` on nested function/module
### 📚 The doc issue

Can we better specify the behavior and eventually the best practices when decorating a function or compiling a module and the effect on the nested modules and nested function call?

https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html

### Suggest a potential alternative/fix

_No response_

cc @sekyondaMeta @svekars @kit1980 @williamwen42 @msaroufim @ezyang @bdhirsh @anijain2305 @zou3519 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### PR DESCRIPTION
**Pull request by ignaciobartol**
Fixes #2858

## Description
Based on the discussion on #2858 I added a few examples on how to use `torch.compile` in nested modules and functions. Also, I added a few lines for "Best practices" to debug if there is a compilation problem and also how to disable a function from being compiled (along with its children functions) using `torch.compiler.disable`.

Let me know if anything can be improved in the tutorial and I will work on it.  As a side note, from this [issue](https://github.com/pytorch/pytorch/issues/123771) in pytorch, I am not sure if `torch.compiler.disable` is working properly, but I used as specified as in [the docs here](https://pytorch.org/docs/stable/torch.compiler_fine_grain_apis.html#torch-compiler-disable)

## Checklist
- [x] The issue that is being fixed is referred in the description (see above "Fixes #ISSUE_NUMBER")
- [x] Only one issue is addressed in this pull request
- [x] Labels from the issue that this PR is fixing are added to this pull request
- [x] No unnecessary issues are included into this pull request.

cc @williamwen42 @msaroufim @sekyondaMeta @svekars @kit1980 @brycebortree

### CODE REVIEW

File: intermediate_source/torch_compile_tutorial.py [hunk 1]
```diff
@@ -94,7 +112,74 @@ def forward(self, x):
...
+except Exception as e:
+    print("Compilation of outer_function failed:", e)
+
+######################################################################
+# Best Practices and Recommendations
+# ----------------------------------
+#
+# Behavior of ``torch.compile`` with Nested Modules and Function Calls
+#
+# When you use ``torch.compile``, the compiler will try to recursively inline
```
Comment (williamwen42) [inline intermediate_source/torch_compile_tutorial.py:162]: Don't have to mention inlining - going over how dynamo inlines is a little more involved.

File: intermediate_source/torch_compile_tutorial.py [hunk 2]
```diff
@@ -94,7 +112,74 @@ def forward(self, x):
...
+# and compile every function call inside the target function or module.
```
Comment (williamwen42) [inline intermediate_source/torch_compile_tutorial.py:163]: "every function call" is not exactly right - there is a skiplist of various functions (e.g. builtins, some functions in the `torch.*` namespace).

**Review by williamwen42 — CHANGES_REQUESTED** [review]
I think this doc update would be better placed in the troubleshooting tutorial (@mlazos is working on this and will have an update soon).

File: intermediate_source/torch_compile_tutorial.py (hunk 2 above)
Comment (ignaciobartol) [reply intermediate_source/torch_compile_tutorial.py:163]: Thanks, so I will modify the wording adding your comment to something like "compile every function call inside the target function or module that is not in a skiplist (e.g. builtins, some functions in the torch.* namespace)."

**Comment by ignaciobartol** [comment]
> I think this doc update would be better placed in the troubleshooting tutorial (@mlazos is working on this and will have an update soon).

Thanks @williamwen42 for the review and the great feedback! I will work on them. Once I have an update on which doc to contribute I will create a new commit.

**Review by williamwen42 — CHANGES_REQUESTED** [review]
Content looks good, but can you fix any formatting issues that you see in the doc preview? 

Also waiting for response from @mlazos to see where this change should go

**Comment by mlazos** [comment]
Thanks @ignaciobartol, this looks like a good tutorial. I think we should add a section called "Best Practices for Onboarding a New Model" to https://pytorch.org/docs/stable/torch.compiler_troubleshooting.html and add your best practices section to that. You can also link to it from your tutorial as well. The rest of the troubleshooting doc can be a bit outdated and I'll be revamping that next week.

**Comment by ignaciobartol** [comment]
> Content looks good, but can you fix any formatting issues that you see in the doc preview?
> 
> Also waiting for response from @mlazos to see where this change should go

Thanks! I am not sure what I am seeing in my build of the docs differs from what is on the preview (See screenshot below). Any ideas of why this may be happening? 

torch_compile_build

File: intermediate_source/torch_compile_tutorial.py [hunk 3]
```diff
@@ -94,7 +98,101 @@ def forward(self, x):
...
+        x = self.inner_module(x)
+        return torch.nn.functional.relu(self.outer_lin(x))
+
+outer_mod = OuterModule()
+opt_outer_mod = torch.compile(outer_mod)
+print(opt_outer_mod(t))
+
+######################################################################
+# We can also disable some functions from being compiled by using
+# `torch.compiler.disable`. Suppose you want to disable the tracing on just
```
Comment (svekars) [inline intermediate_source/torch_compile_tutorial.py:139]:
```suggestion
# ``torch.compiler.disable``. Suppose you want to disable the tracing on just
```

File: intermediate_source/torch_compile_tutorial.py [hunk 4]
```diff
@@ -94,7 +98,101 @@ def forward(self, x):
...
+# the `complex_function` function, but want to continue the tracing back in
```
Comment (svekars) [inline intermediate_source/torch_compile_tutorial.py:140]: Double backticks everywhere - it's different from markdown
```suggestion
# the ``complex_function`` function, but want to continue the tracing back in
```

File: intermediate_source/torch_compile_tutorial.py [hunk 5]
```diff
@@ -94,7 +98,101 @@ def forward(self, x):
...
+# `complex_conjugate`. In this case, you can use
```
Comment (svekars) [inline intermediate_source/torch_compile_tutorial.py:141]:
```suggestion
# ``complex_conjugate``. In this case, you can use
```

File: intermediate_source/torch_compile_tutorial.py [hunk 6]
```diff
@@ -94,7 +98,101 @@ def forward(self, x):
...
+# `torch.compiler.disable(recursive=False)` option. Otherwise, the default is
```
Comment (svekars) [inline intermediate_source/torch_compile_tutorial.py:142]:
```suggestion
# ``torch.compiler.disable(recursive=False)`` option. Otherwise, the default is
```

File: intermediate_source/torch_compile_tutorial.py [hunk 7]
```diff
@@ -94,7 +98,101 @@ def forward(self, x):
...
+# `recursive=True`.
```
Comment (svekars) [inline intermediate_source/torch_compile_tutorial.py:143]:
```suggestion
# ``recursive=True``.
```

File: intermediate_source/torch_compile_tutorial.py [hunk 8]
```diff
@@ -94,7 +98,101 @@ def forward(self, x):
...
+def outer_function(x, y):
+    a = nested_function(x)
+    b = torch.cos(y)
+    return a + b
+
+print(outer_function(t1, t2))
+
+######################################################################
+# In the same fashion, when compiling a module all sub-modules and methods
+# within it, that are not in a skiplist, are also compiled.
```
Comment (svekars) [inline intermediate_source/torch_compile_tutorial.py:121]:
```suggestion
# within it, that are not in a skip list, are also compiled.
```

File: intermediate_source/torch_compile_tutorial.py [hunk 9]
```diff
@@ -94,7 +98,101 @@ def forward(self, x):
...
+
+######################################################################
+# Best Practices and Recommendations
+# ----------------------------------
+#
+# Behavior of ``torch.compile`` with Nested Modules and Function Calls
+#
+# When you use ``torch.compile``, the compiler will try to recursively compile
+# every function call inside the target function or module inside the target
+# function or module that is not in a skiplist (e.g. builtins, some functions in
```
Comment (svekars) [inline intermediate_source/torch_compile_tutorial.py:175]:
```suggestion
# function or module that is not in a skip list (such as built-ins, some functions in
```

**Comment by svekars** [comment]
From publishing perspective the PR looks good to me. @mlazos - did you mean that [this](https://github.com/pytorch/tutorials/pull/2936/files#diff-7f25dcffa2b8860d270f61106b1d5c8ad8a98ad599a3dc7b9e4df9cb281efdf8R168) should be moved to [torch.compiler troubleshooting](https://pytorch.org/docs/stable/torch.compiler_troubleshooting.html) instead of being here?

**Comment by ignaciobartol** [comment]
> From publishing perspective the PR looks good to me. @mlazos - did you mean that [this](https://github.com/pytorch/tutorials/pull/2936/files#diff-7f25dcffa2b8860d270f61106b1d5c8ad8a98ad599a3dc7b9e4df9cb281efdf8R168) should be moved to [torch.compiler troubleshooting](https://pytorch.org/docs/stable/torch.compiler_troubleshooting.html) instead of being here?

Thanks @svekars  for fixing the formatting errors. Once we have more information about this I can ago ahead and apply those changes

**Comment by mlazos** [comment]
> From publishing perspective the PR looks good to me. @mlazos - did you mean that [this](https://github.com/pytorch/tutorials/pull/2936/files#diff-7f25dcffa2b8860d270f61106b1d5c8ad8a98ad599a3dc7b9e4df9cb281efdf8R168) should be moved to [torch.compiler troubleshooting](https://pytorch.org/docs/stable/torch.compiler_troubleshooting.html) instead of being here?

Correct, it was just a suggestion, I'm okay either way.

**Review by williamwen42 — APPROVED** [review]
